### ISSUE TITLE
Create 2 has_many through associations in one Cocoon form

### ISSUE THREAD

**Issue by clveranis** [nathanvda/cocoon#469]
Hey guys,

I've asked this question on [Stack Overflow](https://stackoverflow.com/questions/47216904/create-2-has-many-through-associations-in-one-cocoon-form) first, as recommended because I know GH issues are more for problems with Cocoon rather than how-to-help, but its approaching a week with no responses and in my experience that usually means the question is dead.

Either way, I'm wondering if its possible to tell Cocoon to create a second association on a has_many through relationship. I've broken down the specifics in my SO post. I picked through the core Cocoon code and see that it uses the `create_object_on_association` method which I'm assuming creates both the object as well as the association. So is there a way to make it so I can create a second association with a different model for the same new object I'm creating?

I understand it probably will be a custom method or modifying the current one but I wanted to see if you guys had any ideas so I'm not shooting in the dark. 

Thanks!

**Comment by nathanvda** [comment]
To be honest I saw your question on SO the 11th (I get notified by all cocoon related questions on SO) and you already had found a solution, and your question is very complicated/convulated to read so I just gave up :) 

I am trying to read it again, and it still is very hard to understand. I do not quite understand the problem/question. Did you check the demo application where different relations/associations are demonstrated?

**Comment by clveranis** [comment]
Ah okay, I usually try to provide as much information as possible but I guess its too much. I have looked through the demo app and a regular, single, has_many through works completely fine and what I've done matches what you have and it works as expected.

To keep it as simple as possible, here is the form I use.

    <%= form_for(@orders, url: prop_order_misc_update_patch_path(@properties, @orders), method: "patch") do |f| %>
      <%= render 'shared/error_messages', object: f.object %>
      <%= f.hidden_field :section, value: "propOrderOwner" %>
      <%= f.fields_for :owners do |owner| %>
        <%= render 'properties/owner_fields', f: owner %>
      <% end %>
      <%= link_to_add_association fa_icon("plus", text: "Add Owner"), f, :owners, class: 'btn btn-info btn-outline' %>
      <%= f.form_group do %>
        <%= f.submit "Save Owners", class: "btn btn-primary" %>
      <% end %>
    <% end %>

So this works as expected, creates the `order_owner` association but this `order` belongs_to some `property`, I'd like to add whatever `owners` I'm now creating via this form, to also get created the same way as if this form read:

    form_for(@properties) do |f|
      ...
      f.fields_for :owners do |owners|

**Comment by Mirv** [comment]
So in plain english ... what you want is to create an order for a product.

That order form needs to also have a nested fields for:

order_owner (working)
property (working)
owners (not written? or working?)

I've never asked how to handle this myself, but when filling out multiple relationships to be built off one action, I build them out in reverse, than call save on it in the responds to create.  I'm not exactly sure how this pans out for you, but ...

Imaginary scenario just to demo the flow Company 1-* Person 1-* Order (2 x one 2 many relationships in the models).

(I use before_filters to pass `@company` between actions per michael hartls free online tutorial if necessary)

New action
```
@company = params[:company_id]

# attach lowest level to the mid level
@person1.order1.build
@person2.order2.build

# attach mid level to the upper level
@company.person1 = @person1
@company.person2 = @person2
```

**Comment by clveranis** [comment]
Hey Mirv, thank you for taking the time to comment. 

In plain English, upon creation of a single new object (owners) I want to add it to two associated models (property&order) that already exist at the time of creating the owner.

I understand what you're staying in the demo workflow you typed out, however, it's slightly different than what I'm trying to accomplish. It would be more along the lines of

    ### this is for demo only and is not a literal translation of any code ### 

    @order = Order.find(1)
    @property = Property.find(1)
    (@property.orders << @order) # adding this just to show relationship

    # when filling out an order I add an owner via forms & Cocoon, so lets say step 1:
    @order.owners << owner1

    # and in essence my next step would be, step 2:
    @orders.owners.each do |owner|
      @property.owners << owner
    end

    # so at the end of the whole thing, it would be like this:
    @owner.owners = owner1
    @property.owners = owner1 

So the entire point to all this is just to add a single new record, to two models, the same way as in your demo app would be adding 

    Project <- Project_tags <- Tag1
    Job <- Job_tags <- Tag1

I apologize if this whole thing got convoluted with information overload and if this is not possible to do with Cocoon then I understand. In laymans terms, I don't understand how the method `create_object_on_association` does what it does and I'm just trying to do that x2.

**Comment by Mirv** [comment]
If it's just not understanding what it does ... https://github.com/nathanvda/cocoon/blob/d64a1e3cd2cae7bf4a8f0960a88e4ca5e8697cc4/lib/cocoon/view_helpers.rb#L128

**Comment by Mirv** [comment]
So what is the error when you're doing this?

**Comment by nathanvda** [comment]
Ok, let me try to recap it so it makes sense: 

* order has-many properties
* order has-many owners 

You create a form for an order, right? Yet the instance variable is plural, is that a typo, to confuse the russians or intentional? So if it is plural, why not use the `fields_for` of the parent instead?

But assume it is one order. Then: 

* property has-many owners 
* property has-many orders

Now from the code, the fix, in your SO question I see you add owners to properties if they are owner of an order. This feels an awful lot like duplication to me and I do not really like that. 

Now first off: if that works, and it is a necessity to your project/data-model to have that duplication for whatever reason (caching/performance/...) the solution you found in SO seems perfectly fine. Just to make clear, if it was not obvious before: cocoon is just a helper to make rails nested forms more dynamic, a nested form is a single hierarchy, and not maintaining (what you seem to want) two hierarchies in sync. Of the top of my head you could even use javascript in the form to make sure when an owner is added in one branch, you also add it to the other branch.  

But, suppose you do not like the duplication either, just a crazy idea, but could you not just write: 

* property has-many orders 
* property has-many owners through orders ? 

Would that not fix it, or could a propery have extra owners? Then could you not fix it (because imho it might seem relevant how an owner is related) to have something like 

* property has-many orders 
* property has-many owners 

and add

    def all_owners 
       owners.to_a + orders.map{|order| order.owners.to_a}
    end 

(not tested, not entirely sure that works, but I hope you get the gist of it)

**Comment by clveranis** [comment]
@Mirv Yeah I picked through the view helper code for a few hours and thats how I stumbled upon `create_object_on_association` in the first place and my assumption is thats the method that actually helps create the new object and its association with the parent of the nested field.

@nathanvda Yeah the pluralization `@orders` is somewhat of a typo, to learn rails I worked through Micheal Hartl's Ruby on Rails tutorial and at the time I started this project after that I didn't really understand the convention of single/plural instance variables so now its just woven throughout my app and I have yet to pick through and change it everywhere. 

To answer your modeling questions, its not so much about properties being able to have extra owners as it is that an owner has_many properties, however, there's no need just yet to manage an owner by itself since the focus is on properties and orders for the actual purpose of the app.

Basically,

- property has-many orders
- property has-many owners through orders 
- owner has-many properties

Would be the true modeling of the data because a single owner can have many properties. The way that I have it set up is the best way that I could think of since over time the owner of a property will change, or a property could have multiple owners at once, or the old owner of a property could own other properties but if any of those owners are on an order, they just belong to that instance of an order, etc. 

I fully understand that Cocoon wasn't directed at managing multiple hierarchies but I wasn't sure if it was something easy like "just set up nested forms like this" or "cocoon triggers this built-in rails method so just hook into that" but it doesn't seem to be the case and I apologize for wasting your time on this endeavor. 

My javascript knowledge is only surface level html effects and what not but if using javascript in the way you mentioned is what you yourself might do, then thats the route I'll research and thank you for the suggestion.

**Comment by ktsadmin** [comment]
So before I close this thread, I just had a quick question, is there any way that you know of to make cocoon use a `find_or_create` method when creating a new owner?

Like you said, I know Cocoon is only a helper for `nested_forms` and I've been exhausting all possibilities and searches that I can find to aid with this. From what I gather, when `accepts_nested_attributes_for` saves, it calls the rails `autosave_associated_records_for_*` which can be overwritten, but I can't seem to get it working at all for a has_many :through relationship. Alot of the examples I find and even in your guide you have the look up or create example for belongs_to and I'm trying to do something similar except for a has_many :through.

So I was hoping you had some suggestions or have encountered this before in some way and would have some ideas. Thanks much!

**Comment by clveranis** [comment]
So before I close this thread, I just had a quick question, is there any way that you know of to make cocoon use a find_or_create method when creating a new owner?

Like you said, I know Cocoon is only a helper for nested_forms and I've been exhausting all possibilities and searches that I can find to aid with this. From what I gather, when accepts_nested_attributes_for saves, it calls the rails autosave_associated_records_for_* which can be overwritten, but I can't seem to get it working at all for a has_many :through relationship. Alot of the examples I find and even in your guide you have the look up or create example for belongs_to and I'm trying to do something similar except for a has_many :through.

So I was hoping you had some suggestions or have encountered this before in some way and would have some ideas. Thanks much!

**Comment by nathanvda** [comment]
I do not quite understand. From what you say, you have checked the demo project [here](https://github.com/nathanvda/cocoon_simple_form_demo) where it is shown how to either select an existing person as project-owner, or create a new one, but more appropriately: how to select either an existing tag or create a new one. Which is what you are looking for, right? But to make that work your `through` table has to be explicit. 

So in other words: let the user decide if she wants to or link to an existing object or create a new one. 

Otherwise there is no easy solution. As said before: cocoon only aids on the frontent in building a form more dynamically, there is no server/controller code included, so anything that is not "standard" rails you will have to provide yourself.

**Comment by clveranis** [comment]
Yeah thats the conclusion I came to. I use a plugin called Select2, which I'm sure you've heard of, and that allows you to choose existing records via select box like normal but it also allows you to type into the box to create new records. So I had been trying to figure out how I could make a single select box have the capabilities of both but have hit a wall with the whole `find_or_create` thing. The way you have it in your demo is clearly the easiest way, I just wasn't if you had attempted anything like this so I figured I'd ask.

Thank you for letting me pick your brain and happy holidays.
